nightly-examples/examples/proxy_server.rs

```diff
@@ -0,0 +1,102 @@
+#![feature(async_await)]
+#![deny(rust_2018_idioms)]
+
+//! A proxy that forwards data to another server and forwards that server's
+//! responses back to clients.
+//!
+//! Because the Tokio runtime uses a thread pool, each TCP connection is
+//! processed concurrently with all other TCP connections across multiple
+//! threads.
+//!
+//! You can showcase this by running this in one terminal:
+//!
+//!     cargo +nightly run --example proxy_server
+//!
+//! This in another terminal
+//!
+//!     cargo run --example echo
+//!
+//! And finally this in another terminal
+//!
+//!     nc localhost 8081
+//!
+//! This final terminal will connect to our proxy, which will in turn connect to
+//! the echo server, and you'll be able to see data flowing between them.
+
+use futures::{future::try_join, FutureExt};
+use tokio::{
+    self,
+    io::AsyncReadExt,
+    net::{TcpListener, TcpStream},
+    prelude::*,
+};
+
+use tracing::{debug, info, warn};
+use tracing_attributes::instrument;
+
+use std::{env, net::SocketAddr};
+
+#[instrument]
+async fn transfer(
+    inbound: TcpStream,
+    proxy_addr: SocketAddr,
+) -> Result<(), Box<dyn std::error::Error>> {
+    let outbound = TcpStream::connect(&proxy_addr).await?;
+
+    let (mut ri, mut wi) = inbound.split();
+    let (mut ro, mut wo) = outbound.split();
+
+    let client_to_server = ri.copy(&mut wo);
+    let server_to_client = ro.copy(&mut wi);
```

Comment:
What do you think of giving each of these `copy` futures a span of its own? Like:
```rust
    let client_to_server = ri.copy(&mut wo)
         .map(|bytes| { debug!(message = "copy finished", bytes); bytes })
         .map_err(|error| { warn!(%error); error })
         .instrument(debug_span!("client_to_server"));
    let server_to_client = ro.copy(&mut wi)
         .map(|bytes| { debug!(message = "copy finished", bytes); bytes })
         .map_err(|error| { warn!(%error); error })
         .instrument(debug_span!("server_to_client"));
```

Reply:
I wanted to do this, but I wasn't entirely certain how to go about doing it. Thank you so much for this!

Reply:
I added this, but the only issue is that the trace output from these spans only shows up after either the server or the client disconnect.